---
title: "BCB330 notebook"
output: html_notebook
---
Oct. 21, 2024

Got Esther's code base (/external/rprshnas01/netdata_kcni/dflab/team/eoa/tPRS_analysis/scripts/). She already did tPRS with Maitra et al. (2023)!!!!
TODO: 
* replicate tPRS and PRS for MDD
* separate DEG list from Maitra into cell types
* format sc-cis-eQTLs for DEGs into PredictDB format
* look at ABCD data (go to confluence)

Should I include:
* Serotonergic-like Neurons and Dopaminergic Neurons as ExN and/or InN?
* Astrocyte-like Cells as Ast?
* only Cell-type-specific eQTL (no stimulation conditions) or also Response eQTL?

For now, I will stick to only the exactly labeled sc-eQTls -> then I will only use sc-eQTLs from Bryois et al. (2022)

```{r}
# to download all cis-eQTLs from 
# shell(pip3 install zenodo_get)
# shell(zenodo_get 10.5281/zenodo.7276971)
# I just downloaded one chr per a few cell types
eqtl_dir <- "c:Users/mding/Desktop/BCB330/sc-tprs-mdd/eqtl_bryois"


```

Prediction model format (*.db):
![](extra_table.png)
desc: 
gene - Ensembl ID of the gene

genename - HUGO symbol for the gene (optional)

n_snps_in_window - The number of cis-SNPs examined for model training. This is the number of SNPs found within 1 million base pairs upstream of the transcription start site and 1 million base pairs downstream of the transcription end site. Sites are determined from Gencode v19.

n.snps.in.model - The number of SNPs within the cis window that have non-zero weights, as found by elastic-net. (optional)

(rho_avg - Average correlation between predicted and observed on the hold out folds when doing nested cross-validation (optional))

(rho_zscore - Transformation of rho_avg into a z-score using Stouffer's Method. (optional))

pred.perf.R2 - rho_avg squared (optional)

pred.perf.pval - p-value for rho_zscore (optional)

pred.perf.qval - Deprecated. Previously held q-values calculated from the distribution of p-values, but it was later deemed this analysis was inappropriate. (optional)

![](weight_table.png)
desc:
gene - Ensembl ID of the gene

rsid - rsid of the SNP. rsids are from dbSNP version 150.

varID - String label of the format chromosome_position_allele1_allele2_build. All varIDs should be from build 37 of the Human Reference Genome.

ref_allele - Allele which appears on Reference Genome

eff_allele - Alternative/Effect allele.

weight - The weight for this SNP in predicting expression for the gene. In predicting the expression for the gene, the weight is multiplied by the count, or estimated count, of the effect allele in the individual. This value is added to all other weighted SNPs in the model.

![](abcd_scc_format.png)
* how abcd data is stored in SCC
* /external/rprshnas01/external_data/abcd

Oct. 22, 2024

TODO: 
* replicate tPRS and PRS for MDD
* separate DEG list from Maitra into cell types
* format sc-cis-eQTLs for DEGs into PredictDB format
* look at ABCD data (on SCC)

Talk w/ Esther: 
* where is PRS code?
* thoughts on Maitra et al. (2023), where did it go in the analysis?
* does she follow the numbered R scripts exactly for her overall analysis?
* any idea about the PredictDB file format









Continue web scraping scQTLbase for fun maybe, but all of my eQTLs come from Bryois et al. (2022)...
(look at https://joshuamccrain.com/tutorials/web_scraping_R_selenium.html)
```{r}
library(dplyr)
library(stringr)
library(purrr)
library(rvest)
library(RSelenium)
ct_dict <- c("Astrocytes" = "Ast", "Astrocyte-like Cells" = "Ast", "Endothelial Cell (Endo)" = "End1" #only one with one cluster (regex others)
, "Excitatory Neurons" = "ExN", "Inhibitory Neurons" = "InN", "Microglia (Micro)" = "Mic", "Oligodendrocyte Precursor Cell (OPC)" = "OPC", "Oligodendrocytes" = "Oli" )
library(hash)
rev_ct_dict <- as.list(invert(hash(ct_dict)))
# regex fun str_detect(rev_ct_dict["Ast"], regex("Astrocyte-like Cells", ignore_case = TRUE))

url <- "https://bioinfo.szbl.ac.cn/scQTLbase/Search/"
# html <- read_html(url)
driver <- rsDriver(browser = "firefox",
                   port = 4556L,
                   verbose = FALSE,
                   chromever = NULL)
remDr<- driver[["client"]]
reDr$navigate(url)
# driver[["server"]]$stop()
```

